lib/pkcs11h-openssl-1.1.1.c

```diff
@@ -0,0 +1,1761 @@
+		goto cleanup;
+	}
+
+	/* We want DER encoding of X509_SIG = {algor, digest} which could be
+	 * computed as i2d_X509_SIG(), but, unfortunately, the X509_SIG struct
+	 * is opaque and has no constructor. Hence we combine the two elements
+	 * into a sequence ourselves -- not pretty
+	 */
+
+	/* find required size for the buffer */
+	if ((tmp_len = i2d_X509_ALGOR (algor, NULL)) < 0)
+		goto cleanup;
+	out_len = tmp_len;
+
+	if ((tmp_len = i2d_ASN1_OCTET_STRING (digest, NULL)) < 0)
+		goto cleanup;
+	out_len += tmp_len + 2 ; /* extra 2 bytes for sequence header added below */
+
+	if (_pkcs11h_mem_malloc ((void *)&out, out_len) != CKR_OK) {
+		_PKCS11H_LOG (PKCS11H_LOG_WARN, "PKCS#11: Cannot allocate memory");
+		goto cleanup;
+	}
+
+	ptr = out;
+	*ptr++ = V_ASN1_SEQUENCE | V_ASN1_CONSTRUCTED;
+	*ptr++ = out_len - 2;
+
+	/* compute and append the DER of algor and digest to ptr */
+	i2d_X509_ALGOR(algor, &ptr);  /* this advances ptr */
+	i2d_ASN1_OCTET_STRING(digest, &ptr);
+
+	*enc_len = out_len;
+	*enc = out;
+	out = NULL;
+
+	_PKCS11H_LOG(
+		PKCS11H_LOG_DEBUG1,
+		"PKCS#11: encode_pkcs1 - encoded length = %d", out_len
+	);
+
+cleanup:
+	if (digest)
+		ASN1_STRING_free(digest);
+	if (algor)
+		X509_ALGOR_free(algor);
+	if (out)
+		_pkcs11h_mem_free ((void*)&out);
+
+	return (out_len > 0);
+}
+
+/* return 1 on success -1 on error */
+static int
+__pkcs11h_openssl_pkey_rsa_sign(
+	IN EVP_PKEY_CTX *ctx,
+	OUT unsigned char *sig,
+	IN OUT size_t *siglen,
+	IN const unsigned char *tbs,
+	IN size_t tbslen
+)
+{
+	pkcs11h_certificate_t certificate;
+	PKCS11H_BOOL session_locked = FALSE;
+	CK_RV rv = CKR_FUNCTION_FAILED;
+
+	CK_MECHANISM mech = {CKM_RSA_PKCS, NULL, 0};
+	EVP_PKEY *pkey;
+	CK_RSA_PKCS_PSS_PARAMS pss_params = {0};
+
+	const unsigned char *from = tbs;
+	size_t from_len = tbslen;
+	unsigned char *encoded = NULL;
+	size_t enc_len = 0;
+	int padding;
+	size_t siglen_tmp;
+
+	_PKCS11H_ASSERT (ctx!=NULL);
+	_PKCS11H_ASSERT (sig!=NULL);
+	_PKCS11H_ASSERT (siglen!=NULL);
+	_PKCS11H_ASSERT (tbs!=NULL);
+
+	_PKCS11H_DEBUG (
+		PKCS11H_LOG_DEBUG2,
+		"PKCS#11: __pkcs11h_openssl_pkey_rsa_sign entry - ctx=%p, sig=%p, *siglen="P_Z", tbs=%p, tbslen="P_Z"",
+		(void*)ctx,
+		sig,
+		*siglen,
+		tbs,
+		tbslen
+	);
+
+	EVP_PKEY_CTX_get_rsa_padding (ctx, &padding);
+
+	pkey = EVP_PKEY_CTX_get0_pkey (ctx);
+	if (pkey == NULL) {
+		goto cleanup;
+	}
+
+	switch (padding) {
+		case RSA_PKCS1_PADDING:
+		{
+			/* Add DigestInfo to tbs */
+			EVP_MD *md;
+			int md_type;
+			if (EVP_PKEY_CTX_get_signature_md (ctx, &md) <= 0) {
+				_PKCS11H_LOG (PKCS11H_LOG_WARN, "PKCS#11: unknown signature MD");
+				goto cleanup;
+			}
+
+			md_type = EVP_MD_type(md);
+			if (md_type == NID_md5_sha1) {
+				break; /* DigestInfo wrapper not needed */
+			}
+			else if (!__pkcs11h_encode_pkcs1 (&encoded, &enc_len, md_type, tbs, tbslen)) {
+				_PKCS11H_LOG (PKCS11H_LOG_WARN, "PKCS#11: PKCS1 encoding failed");
+				goto cleanup;
+			}
```

Comment:
can you please explain why openssl does not encode the PKCS#1 by itself? so far this was not needed. it is very strange to me that openssl does not have common logic for this.